config/src/main/java/com/typesafe/config/impl/ConfigParser.java

```diff
@@ -0,0 +1,435 @@
+/**
+ *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
+ */
+package com.typesafe.config.impl;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Stack;
+
+import com.typesafe.config.*;
+
+final class ConfigParser {
+    static AbstractConfigValue parse(ConfigNodeRoot document,
+                                     ConfigOrigin origin, ConfigParseOptions options,
+                                     ConfigIncludeContext includeContext) {
+        ParseContext context = new ParseContext(options.getSyntax(), origin, document,
+                SimpleIncluder.makeFull(options.getIncluder()), includeContext);
+        return context.parse();
+    }
+
+    static private final class ParseContext {
+        private int lineNumber;
+        final private ConfigNodeRoot document;
+        final private FullIncluder includer;
+        final private ConfigIncludeContext includeContext;
+        final private ConfigSyntax flavor;
+        final private ConfigOrigin baseOrigin;
+        final private LinkedList<Path> pathStack;
+        // this is the number of "equals" we are inside,
+        // used to modify the error message to reflect that
+        // someone may think this is .properties format.
+        int equalsCount;
+        // the number of lists we are inside; this is used to detect the "cannot
+        // generate a reference to a list element" problem, and once we fix that
+        // problem we should be able to get rid of this variable.
+        int arrayCount;
+
+        ParseContext(ConfigSyntax flavor, ConfigOrigin origin, ConfigNodeRoot document,
+                FullIncluder includer, ConfigIncludeContext includeContext) {
+            lineNumber = 1;
+            this.document = document;
+            this.flavor = flavor;
+            this.baseOrigin = origin;
+            this.includer = includer;
+            this.includeContext = includeContext;
+            this.pathStack = new LinkedList<Path>();
+            this.equalsCount = 0;
+            this.arrayCount = 0;
+        }
+
+        // merge a bunch of adjacent values into one
+        // value; change unquoted text into a string
+        // value.
+        private AbstractConfigValue parseConcatenation(ConfigNodeConcatenation n) {
+            // this trick is not done in JSON
+            if (flavor == ConfigSyntax.JSON)
+                return null;
```

Comment:
if we're in JSON, shouldn't the DocumentParser have thrown an exception rather than creating a ConfigNodeConcatenation? If so then this `return null` would be impossible (it could be a BugOrBroken instead perhaps)